_posts/2018-05-07-is-graphql-the-future.markdown

```diff
@@ -0,0 +1,196 @@
+* **Object** types (`type`, `union` and `interface`) are collections of fields. As an operation, an object-typed field is an intermediate operation that serves as the junction point for follow-up operations. But itself, it doesn’t return any data. They are the branches of the request tree.
+
+The entire model for a given API is known as its **schema**. Every schema has a root query object type, whose fields serve as the API’s entry points.
+
+```graphql
+# The root query object type
+type Query {
+  artwork(id: ID): Artwork
+  artist(name: String)
+  # … a whole bunch more root fields
+}
+
+type Artwork {
+  title: String
+  artist: Artist
+}
+
+type Artist {
+  name: String
+}
+```
+
+A GraphQL request begins by mentioning at least one of the fields of the root query object. This represents an initial query. And if that field is an object, _its_ fields any number of follow-up queries. Critically, _any_ field in the request tree can take arguments, allowing a request to be parameterized at all depths.
+
+Take this query, for example:
+
+```graphql
+{
+  artwork(id: "andy-warhol-campbells-soup-i-black-bean") {
+    title
+    artist {
+      name
+    }
+  }
+}
+```
+
+Here, we tell the server to look up an Artwork by its slug, and tell us the title. So far, this is just like REST. But we _also_ tell it to find us the Artist for us. Importantly, object fields _must_ be followed up with further queries, and scalar fields _cannot_ be. With that in mind, it’s easy to see that `artwork` and `artist` are object fields, while `title` and `name` are scalar fields.
+
+Also note that the fact that there’s also an `artist` root query field actually has nothing to do with its presence under `Artwork`. There can be multiple paths to reach the same GraphQL type. This is defined explicitly by the schema.
+
+Usefully, the server’s response to a GraphQL request will directly mirror the shape of the request itself. The result of the query above looks like:
+
+```
+{
+  "data": {
+    "artwork": {
+      "title": "Campbell's Soup I: Black Bean",
+      "artist": {
+        "name": "Andy Warhol"
+      }
+    }
+  }
+}
+```
+
+# GraphQL as a (meta-)scripting language
+
+Let’s dig a little deeper into the scripting language interpretation of GraphQL, because this is the crux of how I think people should think of GraphQL. If I were to guess, I think Facebook…
+
+* …knows this is true. After all, much of the spec is devoted to [the execution model of GraphQL](http://facebook.github.io/graphql/October2016/#sec-Execution).
+* …might have backed into this design. It’s well known that they think of their data as a graph, so I suspect GraphQL might have begun literally as a graph query language.
```

Comment:
Would be cool if `graph query language` had a link out to something like https://developer.ibm.com/dwblog/2017/overview-graph-database-query-languages/, just so the idea is clear

Reply:
👍